picoquic/frames.c

```diff
@@ -933,44 +932,73 @@ static int picoquic_queue_network_input(picoquic_cnx_t* cnx, picoquic_stream_hea
         }
     }
 
-    /* Queue of a block in the stream */
-
-    while (next != NULL && start < length && next->offset <= offset + start) {
-        if (offset + length <= next->offset + next->length) {
-            start = length;
-        } else if (offset < next->offset + next->length) {
-            start = (size_t)(next->offset + next->length - offset);
-        }
-        pprevious = &next->next_stream_data;
-        next = next->next_stream_data;
-    }
-
     if (start < length) {
-        size_t data_length = length - start;
+        size_t data_length = 0;
 
-        if (next != NULL && next->offset < offset + length) {
-            data_length -= (size_t)(offset + length - next->offset);
+        /* check for data that is already received in blocks with offset <= offset + start */
+        memset(&target, 0, sizeof(picoquic_stream_data_node_t));
+        target.offset = offset + start;
+        previous = (picoquic_stream_data_node_t*)picosplay_find_previous(&stream->stream_data_tree, &target);
+        if (previous == NULL) {
+            next = (picoquic_stream_data_node_t*)picosplay_first(&stream->stream_data_tree);
+        }
+        else {
+            next = (picoquic_stream_data_node_t*)picosplay_next(&previous->stream_data_node);
+            /* By definition, previous->offset <= offset. Check whether the
+             * beginning of the frame is already received */
+            if (offset + start < previous->offset + previous->length) {
+                if (offset + length <= previous->offset + previous->length) {
+                    start = length;
+                }
+                else {
+                    start = (size_t)(previous->offset + previous->length - offset);
+                }
+            }
         }
 
-        if (data_length > 0) {
-            picoquic_stream_data_t* data = (picoquic_stream_data_t*)malloc(sizeof(picoquic_stream_data_t));
+        /* Check whether parts of the frame are covered by the next frames */
+        while (ret == 0 && start < length) {
+            data_length = length - start;
 
-            if (data == NULL) {
-                ret = picoquic_connection_error(cnx, PICOQUIC_ERROR_MEMORY, 0);
+            if (next != NULL && next->offset < offset + length) {
```

Comment:
When I understand correctly, this may lead to unintended data loss. Consider the following example:

A block of 6 bytes is received at offset 2. And there is already another block of 2 bytes received at offset 4:

```
           0 1 2 3 4 5 6 7 8 9
available:         xxxx
received:      xxxxxxxxxxxx
```

With this condition you would threw bytes 4 to 7, from which bytes 6 and 7 haven't been received yet.

Or maybe I just don't understand :)

Reply:
Ah, ok. Turns out, I didn't understand. Since there is a while loop around this condition, it eventually also handles bytes 6 and 7 mentioned above in a later iteration.

Reply:
Sorry that this is hard to read. I will check that we have a test case covering the overlap scenario.

Reply:
Good thing you flagged that, there was in fact a bug and the new test found it.